### PR TITLE
Expectation function

### PR DESCRIPTION
**Pull request by MrRobot2211**
Possible implementations of the CuPyDense expectation function

### CODE REVIEW

**Comment by MrRobot2211** [comment]
GPU times State ket expectation
GPU times State ket expectation

This are the timings obtained for the different implementation for the operator ket expectation. The functions listed here where used with the caveat that I removed shape check for all as well as the call to item to speed up the trials.

**Comment by MrRobot2211** [comment]
And I bring good news on the kernel vs cupy for density matrices.
gpu times expectation on a dm

3 times speed-up 🎉

**Comment by MrRobot2211** [comment]
There may be more improvements if for example we are okey assuming the operator is hermitian  (I assume we are making a measurement here) .  And we could also use the hermitian property of density matrices.

**Comment by hodgestar** [comment]
Let's pick the best implementation, remove the others and clean up the code and uncomment the tests and then I can do a final review.

File: src/qutip_cupy/dense_functions.py
Comment (hodgestar) [inline src/qutip_cupy/dense_functions.py:4]: This import should stay where it was up top (it's from a third-party package).

File: src/qutip_cupy/expectation.py
Comment (hodgestar) [inline src/qutip_cupy/expectation.py:39]: Could we name this `expect_dense` and rename `x1` to `op` and `x2` to `dm` or `state` to make everything clearer?

File: src/qutip_cupy/expectation.py
Comment (hodgestar) [inline src/qutip_cupy/expectation.py:57]: The execution speed is probably approximately linear in the number of blocks so we really should make the block size as high as possible. Since 2010, 1024 is allowed, so I suggest we pick that.

File: src/qutip_cupy/expectation.py
Comment (hodgestar) [inline src/qutip_cupy/expectation.py:56]: I think we can just remove the TODO (or alternatively enhance it to explain what is wrong with the current settings).

**Review by hodgestar — CHANGES_REQUESTED** [review]
I left some comments and requests for changes. I also noticed that the specialisation is again not registered, so please do that.

Once those changes are made, I am happy for you to merge this without further approval if you need to for GSoC (but please double check yourself in this case!)

File: src/qutip_cupy/expectation.py
Comment (MrRobot2211) [reply src/qutip_cupy/expectation.py:57]: The issue with 1024, is that it is the cap for many GPUs and you may end up with a problem if say for example you have only one GPU and you are running graphics there.

Comment (hodgestar) [reply src/qutip_cupy/expectation.py:57]: Hmm. Perhaps let's just document that it a comment for now then?

Comment (MrRobot2211) [reply src/qutip_cupy/expectation.py:57]: To be clear what actually happens is that using 1024 may take up too many registers and too much shared memory and not let a block from another process run in the same SM. But I'll leave it a t 1024 and leave the comment just in case.

**Review by hodgestar — APPROVED** [review]
Approved!

**Pull request merged**
